### ISSUE TITLE
Extract features of board with trained model

### ISSUE THREAD

**Issue by fallcat** [lightvector/KataGo#359]
Hello! Thank you so much for all the great work! I am working on a project and I hope to use some intermediate layer of your trained model. I saw that you have the code `python/genboard_run.py` in pytorch. It seems to be different from `python/play.py` which uses tensorflow instead. The former doesn't seem to have a policy network and a value network? 

I'm wondering 
1. What is the best way if I want to load the model (in python) and use like for example the penultimate layer.
2. Should I use the policy network or value network or both?
3. How does the input look like?

Thank you in advance!

**Comment by lightvector** [comment]
Genboard is purely a separate side experiment to try to generate board positions, you should ignore it, it has nothing to do with KataGo's normal engine or network or anything.

1. Download Tensorflow weights from https://d3dndmfyhecmj0.cloudfront.net/g170/index.html and use `python/play.py`, or `python/test.py` as an example - you can see how the model is invoked in `fetch_output`, and how the input features are computed using `model.fill_row_features`. The model itself is implemented in python/model.py. 

You should expect to have to read the code and do a bit of work yourself to understand it, since unfortunately, the python code isn't very well documented. `play.py` is also not perfectly-well maintained since all normal users use the C++ engine and nobody uses this, so it's possible the code has gotten slightly stale. (Casual users simply using KataGo and not wanting to peek inside the net should just use the C++ engine, which has its own implementation completely separate from the python).

2. There is only one network, they are not separate networks. This one networks has *many* output heads, not just value and policy. Check out KataGo's paper for a description of the neural net architecture and some of the output heads and input features https://arxiv.org/abs/1902.10565 and you can see how it matches with implementation in model.py. (more output heads have since been added as well). Also, if you're curious what the whole masking business is, check out https://github.com/lightvector/KataGo/blob/master/docs/KataGoMethods.md#training-on-multiple-board-sizes-via-masking

3. Answered above.
 
Good luck! If you discover or build anything interesting regarding the intermediate layers, I'd love to hear about it.

**Comment by lightvector** [comment]
Also, you can check out `cpp/neuralnet/nninputs.cpp` `NNInputs::fillRowV7` for the C++ implementation of the input features, if that also helps - it should be doing the same thing as the python.

And you can check out `python/visualize.py` or `python/export_model.py` to see examples of code that don't actually evaluate the neural net, but rather extract and visualize and export the *raw* weight matrices (variables) within the model.

**Comment by lightvector** [comment]
Lastly, keep in mind that the Tensorflow weights are under a name scope "swa_model", due to being exported using stochastic weight averaging. You might have to specify that name scope when loading a model. You should also see it plainly prefixing all the variable names if you use any of TF1.5's low-level utilities for examining directly inside the savedmodel checkpoints.

**Comment by fallcat** [comment]
Thank you so much for the detailed reply! It makes more sense to me now. I'll read your paper and code and see how to use it.

My project is to determine whether a board and a comment are aligned. I'll let you know if it works out.
